src/compiler/binder.ts

```diff
@@ -278,6 +278,16 @@ namespace ts {
                     Debug.assert(isWellKnownSymbolSyntactically(nameExpression));
                     return getPropertyNameForKnownSymbolName(idText((<PropertyAccessExpression>nameExpression).name));
                 }
+                if (isPrivateName(name)) {
+                    // containingClass exists because private names only allowed inside classes
+                    const containingClass = getContainingClass(name.parent);
+                    if (!containingClass) {
+                        // we're in a case where there's a private name outside a class (invalid)
+                        return undefined;

```

Comment:
I see that this is the only explicit return value of `undefined`. Any other `undefined` would have to come from an inner method call. I see some existing calling code that does things like `!getDeclarationName(...)`. Does this `undefined` value here mean the same thing as any other potential `undefined` values? If not, maybe it's better to throw here (explicitly or implicitly) instead?

Reply:
@Neuroboy23 these changes were reviewed internally months ago and also given a peruse by Daniel and Ron

Reply:
Re this specific question:

- a private name outside a class body is a parse error and will be caught by the parser. If the user has noEmitOnError: false in their compilerOptions then we have to do best-effort on the emit, even though the emit almost certainly bad. In this case, I'm not sure whether it would be better to return `node` or `undefined`.